### PR TITLE
Add backwards compatible CRD versioning

### LINKED ISSUE #215: Generate client with code-generator
The [code generator](https://github.com/kubernetes/code-generator) can generate a client for us.

This will allow us to:

* Conform more easily to the expected structure of other controllers/operator
* Support multiple versions without having to manually rewrite/modify the existing client
* Write less code!

### LINKED ISSUE #216: Add habv1beta2 version
We decided to support the existing `Deployment`-based behavior in `habv1beta`, and to move the `StatefulSet`-based logic to `habv1beta2`.

This allows us to gracefully deprecate the old behavior without disrupting existing clusters.

Additionally, we decided to keep supporting `habv1beta1` for 2 releases,

See #215.

### PR DESCRIPTION
**Pull request by asymmetric**
This PR does several things.

## CustomVersion

There is currently no way to run multiple versions of a CRD. The `CustomVersion` field works around this by "overloading" the Habitat type.

All controllers receive all `Habitat` objects, but then decide which ones to skip based on the value (and presence) of the `CustomVersion` field.

## Code generator

This PR also reorganizes the `pkg` tree to match the expectations of the code-generator. The code-generator is used to generate client, listers and watchers.

In the future, we will be able to maintain multiple versions more easily thanks to this structure.

The use of the code-generator also implied changes in the controller, because the structure of the client, listers and watchers has changed.

## Backwards compatibility

The new field allows us to maintain backwards-compatibility with clusters that are already running `Deployment`-backed `Habitat` objects.

Closes #215.
Closes #216.

### CODE REVIEW

File: cmd/habitat-operator/main.go [hunk 1]
```diff
@@ -122,6 +122,68 @@ func run() int {
...
+		level.Info(logger).Log("msg", "Habitat CRD already exists, continuing")
+	} else {
+		level.Info(logger).Log("msg", "created Habitat CRD")
+	}
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(cSets.KubeClientset, resyncPeriod)
+	habInformerFactory := habinformers.NewSharedInformerFactory(cSets.HabClientset, resyncPeriod)
+
+	config := habv1beta2controller.Config{
+		HabitatClient:          cSets.HabClientset.HabitatV1beta1().RESTClient(),
```
[the remark this answers is not in the document]
Comment (asymmetric) [reply cmd/habitat-operator/main.go]: If we change the v1beta1 version, we introduce a backwards incompatible change, unless we decide to deal with 3 versions (`v1beta1`, `v1betacustom1` and `v1betacustom2`).

I'd say we should just stick to this scheme.

File: pkg/controller/v1beta2/controller.go [hunk 2]
```diff
@@ -147,20 +153,7 @@ func (hc *HabitatController) Run(workers int, ctx context.Context) error {
...
-
-	hc.habInformer = cache.NewSharedIndexInformer(
-		source,
-
-		// The object type.
-		&habv1beta1.Habitat{},
-		resyncPeriod,
-		cache.Indexers{},
-	)
+	hc.habInformer = hc.config.HabitatInformerFactory.Habitat().V1beta1().Habitats().Informer()
```
[the remark this answers is not in the document]
Comment (asymmetric) [reply pkg/controller/v1beta2/controller.go]: Yeah, I feel that maybe we should find a better place to explain the whole thing.

OTOH, it's hopefully temporary so not sure if it's worth it.

File: test/e2e/resources/persisted/habitat.yml [hunk 3]
```diff
@@ -1,16 +0,0 @@
-apiVersion: habitat.sh/v1beta1
```
[the remark this answers is not in the document]
Comment (asymmetric) [reply test/e2e/resources/persisted/habitat.yml]: It now lives under `test/e2e/v1beta1/resources/persisted/habitat.yml`

[the remark this answers is not in the document]
Comment (asymmetric) [reply test/e2e/resources/persisted/habitat.yml]: Good point, we were not catching this because we don't actually run that test.

**Comment by asymmetric** [comment]
Re: your comments on `specv1beta2`: I am thinking we can avoid making backwards-incompatible type changes for now, and wait for proper support for multiple versions to land.

Once that happens, we will simply remove the `customVersion` field, and checks for it, without having to change much else.

If we introduce a completely separate field for the spec, once support for multiple versions lands, we will have to support looking for two different fields (`spec` and `specv1beta2`), and also handle conflicts (what happens if a type has both `spec` and `specv1beta2`?)

**Comment by krnowak** [comment]
> Re: your comments on `specv1beta2`: I am thinking we can avoid making backwards-incompatible type changes for now, and wait for proper support for multiple versions to land.

Making backward-incompatible type changes is not a thing we plan to do, yes, but you never know. Anyway, I was opting for a separate spec struct per version, so we don't have a struct with bag of fields that are supported in various versions. `spec` would contain fields supported only in `v1beta1`, `specv1beta2` for `v1beta2` and so on, so it would be all clear. It hints redundancy, but considering that all the supported versions but last one are effectively "frozen", redundancy should not be a problem.

Just want to avoid having someone using the `v1beta1` custom version and complaining that persistence doesn't work. It wouldn't work, because the operator silently ignored it. And modifying `v1beta1` operator to learn about fields specific to `v1beta2` just to warn about them being ignored does not sound like a thing that will work in long run.

> Once that happens, we will simply remove the `customVersion` field, and checks for it, without having to change much else.

Who's "we"? Developers? We will need to support it anyway for some time, even if we release a proper `v1betaX` version via CRD. And probably introducing a separate CRD version will result in a duplication of the `spec` struct anyway.

If "we" are users, then the change will involve renaming `specv1beta2` to `spec`, removing `customVersion` field and updating the `apiVersion` field.

> If we introduce a completely separate field for the spec, once support for multiple versions lands, we will have to support looking for two different fields (`spec` and `specv1beta2`), and also handle conflicts (what happens if a type has both `spec` and `specv1beta2`?)

With introduction of `v1beta2` customVersion I would consider `v1beta1` frozen. So nothing should be added for the old version. And having a separate `spec` struct per version would reflect that. About conflicts - we still have the `customVersion` field and I want to keep it. This field will decide which operator will handle the manifest and which `spec` struct to expect. So if there is `customVersion` set to `v1beta1` and there is only `specv1beta2` field, then the operator will bail out, because `spec` is missing. If for the same `customVersion` actually both `spec` and `specv1beta2` are specified in the manifest then I would say that the old operator should just ignore the `specv1beta2` field and use `spec` - the operator has no clue about `specv1beta2` at all. And the new operator will ignore the manifest because `customVersion` is different from what it supports.

**Comment by asymmetric** [comment]
In practice this would look like:

```yaml
apiVersion: habitat.sh/v1beta1
kind: Habitat
metadata:
  name: example-persistent-habitat
customVersion: v1beta2
spec:
  # the core/redis habitat service packaged as a Docker image
  image: kinvolk/redis-hab
  count: 3
  # the presence of this key activates persistent storage
  v1beta2:
    persistentStorage:
      # the size of the volume that will be mounted in each Pod
      size: 1Gi
      # a StorageClass object with name "standard" must be created beforehand by
      # the cluster administrator
      storageClassName: example-sc
      # the location under which the volume will be mounted
      mountPath: /tmp/foobar
  env:
    - name: HAB_REDIS
      # this is needed to make redis accept connections from other hosts
      # see https://redis.io/topics/security#protected-mode
      # NOTE: do not use this setting in a production environment
      value: '{ "protected-mode": "no" }'
  service:
    name: redis
    topology: leader
    # if not present, defaults to "default"
    group: foobar
```
      # this is needed to make redis accept connections from other hosts
      # see https://redis.io/topics/security#protected-mode
      # NOTE: do not use this setting in a production environment
      value: '{ "protected-mode": "no" }'
  service:
    name: redis
    topology: leader
    # if not present, defaults to "default"
    group: foobar
```

**Comment by krnowak** [comment]
> So to summarize, your wish to use `specv1beta2` is to avoid users creating a manifest with `customVersion = v1beta1`, but then expecting e.g. persistence.
> 
> Couldn't we solve that with documentation?

We could solve it with documentation, but how would documentation look like if we have a single `spec` thing? "This field is available only in v1beta2", "This field has a bit different behaviour in v1beta3", "This field is optional starting from v1beta2". Kind of a mess, I think. Would be better to document all the fields separately for each version. Also, I think it is better if the operator tells you what you did wrong openly and that maybe will make you go to see the docs.

> Having both customVersion and specv1beta2 seems unnecessary to me, because once customVersion is set, you don't really need `specv1beta2` anymore.

The `specv1beta2` name is only because we can't use the `spec` name again. I mean, we could if we had something like this:

```
type SpecV1Beta1 {
    …
}

type SpecV1Beta2 {
    …
}

type Habitat struct {
    // usual k8s fields, versions, meta, blah blah blah
    CustomVersion string
    Spec json.RawMessage // parsed as SpecV1Beta1 or SpecV1Beta2, depending on CustomVersion
}
```

But I'm not really sure how that flies with yaml and client-go and whatnot. So I thought that repeating custom version in spec name is the least bad thing (better than something like `spec2`, for example).

> Could we e.g. add a subkey to the spec instead, like spec.v1beta2? That would be cleaner IMO, and would also solve the documentation problem.

This `spec.v1beta2` would a be struct that `specv1beta2` was meant to be originally, right? If so, then yeah, it is also a good idea, I think.

**Comment by krnowak** [comment]
Ah, now it's clear what you meant by `spec.v1beta2`. Well, that won't save us from duplication, because if we have `v1beta3` we will need to duplicate all the persistence stuff there anyway. And if I will want to add new things to `service`, where do I put it? `spec.v1beta2.service.newstuff` or `spec.service.v1beta2.newstuff`?

I rather imagined something like:
```
apiVersion: habitat.sh/v1beta1
kind: Habitat
metadata:
  name: example-persistent-habitat
customVersion: v1beta1
spec:
  image: kinvolk/redis-hab
  count: 3
  env:
    - name: HAB_REDIS
      value: '{ "protected-mode": "no" }'
  service:
    name: redis
    topology: leader
    group: foobar
```

and

```
apiVersion: habitat.sh/v1beta1
kind: Habitat
metadata:
  name: example-persistent-habitat
customVersion: v1beta2
spec:
  v1beta2:
    image: kinvolk/redis-hab
    count: 3
    persistentStorage:
      size: 1Gi
      storageClassName: example-sc
      mountPath: /tmp/foobar
    env:
      - name: HAB_REDIS
      value: '{ "protected-mode": "no" }'
    service:
      name: redis
      topology: leader
      group: foobar
```

**Comment by asymmetric** [comment]
> if we have v1beta3 we will need to duplicate all the persistence stuff there anyway

In a hypothetical `v1beta3`, you would have:

```
apiVersion: habitat.sh/v1beta1
kind: Habitat
metadata:
  name: example-persistent-habitat
customVersion: v1beta2
spec:
  image: kinvolk/redis-hab
  count: 3
  v1beta2:
    persistentStorage:
      size: 1Gi
      storageClassName: example-sc
      mountPath: /tmp/foobar
  v1beta3:
    foo:
      bar: baz
  env:
    - name: HAB_REDIS
      value: '{ "protected-mode": "no" }'
  service:
    name: redis
    topology: leader
    # if not present, defaults to "default"
    group: foobar
```

i.e. each field would be related to a release. Is it ideal? No. But we are working around a deficiency in Kubernetes.

> How would documentation look like if we have a single spec thing? "This field is available only in v1beta2", "This field has a bit different behaviour in v1beta3", "This field is optional starting from v1beta2". Kind of a mess, I think. Would be better to document all the fields separately for each version.

This is a good point.

**Comment by krnowak** [comment]
About hypothetical `v1beta3` - this repeats the thing we had with `spec` having fields for all the versions. Now `v1beta2.persistentStorage` will be interpreted by `v1beta3` operator, which may be something different from `v1beta2` operator's interpretation… Also, you have made a mistake there, and the example would deploy `v1beta2` version of the object silently ignoring `v1beta3` field, because you wrote `customVersion: v1beta2`.

About CRD versioning: This is for the future, but I suppose that when CRD versioning lands, we don't do anything immediately I think. I would say that if we need to make another breaking change, then we switch to CRD versioning at the same time. So if the latest custom version is `v1betaN` then the new, breaking CRD version could be `v1betaN+1` without all that custom version and spec hacks. Over time we can deprecate and phase out the oldest versions, which will eventually make custom version to completely fade away.

**Comment by asymmetric** [comment]
PTALA.

File: pkg/controller/v1beta2/utils.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+type keyNotFoundError struct {
+	key string
+}
+
+func (err keyNotFoundError) Error() string {
+	return fmt.Sprintf("could not find Object with key %s in the cache", err.key)
+}
+
+func validateCustomObject(h habv1beta1.Habitat) error {
+	spec := h.Spec.V1beta2
```
Comment (krnowak) [inline pkg/controller/v1beta2/utils.go]: I assume that this is the first time we access the fields of `spec.V1beta2`, so a `nil` pointer check should be in place here, right?

File: pkg/controller/v1beta2/stateful_sets.go [hunk 5]
```diff
@@ -290,48 +281,48 @@ func (hc *HabitatController) cacheStatefulSets() {
...
 
-	h, err := hc.getHabitatFromLabeledResource(d)
+	h, err := hc.getHabitatFromLabeledResource(sts)
 	if err != nil {
-		level.Error(hc.logger).Log("msg", "Could not find Habitat for StatefulSet", "name", d.Name)
+		level.Error(hc.logger).Log("msg", "Could not find Habitat for StatefulSet", "name", sts.Name)
 		return
 	}
 
 	hc.enqueue(h)
```
Comment (krnowak) [inline pkg/controller/v1beta2/stateful_sets.go]: Call to the `checkCustomVersionMatch` function is missing here. We don't want the `v1beta2` operator to handle `v1beta3` stateful sets. I'm wondering if this `checkCustomVersionMatch` could be added to the `enqueue` function to have a single place where this thing is checked.

File: pkg/controller/v1beta2/stateful_sets.go [hunk 6]
```diff
@@ -290,48 +281,48 @@ func (hc *HabitatController) cacheStatefulSets() {
...
```
Comment (krnowak) [inline pkg/controller/v1beta2/stateful_sets.go]: Ditto.

File: pkg/controller/v1beta2/stateful_sets.go [hunk 7]
```diff
@@ -290,48 +281,48 @@ func (hc *HabitatController) cacheStatefulSets() {
...
 	}
 
-	h, err := hc.getHabitatFromLabeledResource(d)
+	h, err := hc.getHabitatFromLabeledResource(sts)
 	if err != nil {
 		// Could not find Habitat, it must have already been removed.
-		level.Debug(hc.logger).Log("msg", "Could not find Habitat for StatefulSet", "name", d.Name)
+		level.Debug(hc.logger).Log("msg", "Could not find Habitat for StatefulSet", "name", sts.Name)
 		return
 	}
```
Comment (krnowak) [inline pkg/controller/v1beta2/stateful_sets.go]: Ditto.

File: pkg/apis/habitat/v1beta1/types.go [hunk 8]
```diff
@@ -51,8 +60,9 @@ type HabitatSpec struct {
 	// The EnvVar type is documented at https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#envvar-v1-core.
 	// Optional.
 	Env []corev1.EnvVar `json:"env,omitempty"`
-	// Optional.
-	PersistentStorage *PersistentStorage `json:"persistentStorage,omitempty"`
+	// +optional
+	// V1beta2 are fields for the v1beta2 type.
+	V1beta2 *V1beta2 `json:"v1beta2"`
```
Comment (krnowak) [inline pkg/apis/habitat/v1beta1/types.go]: Capitalize `b` in `beta` so we have `V1Beta2`? WDYT?

File: pkg/apis/habitat/v1beta1/types.go [hunk 9]
```diff
@@ -61,7 +61,8 @@ type HabitatSpec struct {
 	// Optional.
 	Env []corev1.EnvVar `json:"env,omitempty"`
 	// +optional
```
[the remark this answers is not in the document]
Comment (krnowak) [reply pkg/apis/habitat/v1beta1/types.go]: I made another comment about it, later. This one was made on a commit and github marked it as outdated immediately.

File: pkg/apis/habitat/v1beta1/types.go [hunk 10]
```diff
@@ -61,7 +61,8 @@ type HabitatSpec struct {
...
-	PersistentStorage *PersistentStorage `json:"persistentStorage,omitempty"`
+	// V1beta2 are fields for the v1beta2 type.
+	V1beta2 *V1beta2 `json:"v1beta2"`
```
[the remark this answers is not in the document]
Comment (krnowak) [reply pkg/apis/habitat/v1beta1/types.go]: I made another comment about it, later. This one was made on a commit and github marked it as outdated immediately.

File: Makefile [hunk 11]
```diff
@@ -18,15 +18,15 @@ image: linux
 	$(SUDO) docker build -t "$(IMAGE):$(TAG)" .
 
 test:
-	go test -v $(shell go list ./... | grep -v /vendor/ | grep -v /test/)
+	go test $(shell go list ./... | grep -v /vendor/ | grep -v /test/)
```
[the remark this answers is not in the document]
Comment (asymmetric) [reply Makefile]: You're right that the commit message is maybe not very descriptive, but the point is that these are the unit tests, and the output was [basically useless](https://travis-ci.org/habitat-sh/habitat-operator/builds/365094380#L528) IMO.

Where we don't see any output for a very long time is in e2e tests.

File: pkg/controller/v1beta2/utils.go (hunk 4 above)
Comment (asymmetric) [reply pkg/controller/v1beta2/utils.go]: Good catch. Hopefully we can introduce schema validation soon.

File: pkg/controller/v1beta2/stateful_sets.go (hunk 5 above)
Comment (asymmetric) [reply pkg/controller/v1beta2/stateful_sets.go]: Good point, I'll do that.

**Comment by asymmetric** [comment]
PTALA.

File: Makefile (hunk 11 above)
[the remark this answers is not in the document]
Comment (krnowak) [reply Makefile]: Oh, right. Those are unit tests.

No output in e2e tests may be caused by the fact the the output from the test case is printed when the test case finishes. So if the test case takes long to complete, you won't see anything for a long time.

**Comment by krnowak** [comment]
LFAD.

**Comment by asymmetric** [comment]
OMG.

**Pull request merged**